File: 542_01_Matrix.cpp
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& matrix) {
        if (matrix.size() == 0 || matrix[0].size() == 0)
            return matrix;
        int m = matrix.size(), n = matrix[0].size();
        int maximum = m + n;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (matrix[i][j] > 0) {
                    int up = i > 0 ? matrix[i - 1][j] : maximum;
                    int left = j > 0 ? matrix[i][j - 1] : maximum;
                    matrix[i][j] = min(up, left) + 1;
                }
            }
        }
        for (int i = m - 1; i >= 0; --i) {
            for (int j = n - 1; j >= 0; --j) {
                if (matrix[i][j] > 0) {
                    int down = i < m - 1 ? matrix[i + 1][j] : maximum;
                    int right = j < n - 1 ? matrix[i][j + 1] : maximum;
                    matrix[i][j] = min(min(down, right) + 1, matrix[i][j]);
                }
            }
        }
        return matrix;
    }
};

File: 542_01_Matrix.py
from collections import deque
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        # BFS
        m, n = len(mat), len(mat[0])
        q = deque([])
        for i in range(m):
            for j in range(n):
                if mat[i][j] == 0:
                    q.append((i, j))
                else:
                    mat[i][j] = -1
        
        while q:
            i, j = q.popleft()
            for di, dj in [(-1, 0), (0, -1), (1, 0), (0, 1)]:
                ni, nj = i + di, j + dj
                if 0 <= ni < m and 0 <= nj < n and mat[ni][nj] == -1:
                    mat[ni][nj] = mat[i][j] + 1
                    q.append((ni, nj))
        
        return mat


class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        # DP
        m, n = len(mat), len(mat[0])
        for i in range(m):
            for j in range(n):
                if mat[i][j] == 0:
                    continue
                up = inf if i == 0 else mat[i - 1][j]
                left = inf if j == 0 else mat[i][j - 1]
                mat[i][j] = min(up, left) + 1
        
        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                if mat[i][j] == 0:
                    continue
                down = inf if i == m - 1 else mat[i + 1][j]
                right = inf if j == n - 1 else mat[i][j + 1]
                mat[i][j] = min(mat[i][j], down + 1, right + 1)
        
        return mat
